text/0000-std-rpc-api.md

```diff
@@ -0,0 +1,393 @@
+
+```cs
+[ClientRpc]
+void PongClientRpc(int framekey) { /* ... */ }
+```
+
+Developer can invoke a `ClientRpc` by making a direct function call with parameters:
+
+```cs
+void Update()
+{
+    if (Input.GetKeyDown(KeyCode.P))
+    {
+        PongClientRpc(Time.frameCount); // Server -> Client
+    }
+}
+```
+
+Marking method with `[ClientRpc]` and putting `ClientRpc` suffix to the method name are required, otherwise it will prompt error messages:
+
+```cs
+// Error: Invalid, missing 'ClientRpc' suffix in the method name
+[ClientRpc]
+void Pong(int framekey) { /* ... */ }
+
+// Error: Invalid, missing [ClientRpc] attribute on the method
+void PongClientRpc(int framekey) { /* ... */ }
+```
+
+`[ClientRpc]` attribute and matching `...ClientRpc` suffix in the method name are there to make it crystal clear for RPC call sites to know when they are executing an RPC, it will be replicated and executed on the client-side, without necessarily jumping into original RPC method declaration to find out if it was an RPC, if so whether it is a ServerRpc or ClientRpc:
+
+```cs
+Pong(framekey); // Is this an RPC call?
+
+PongRpc(framekey); // Is this a ServerRpc call or ClientRpc call?
+
+PongClientRpc(framekey); // This is clearly a ClientRpc call
+```
+
+### Reliability
+
+RPCs are reliable by default which means they are guaranteed to be executed on the remote side. However, sometimes developer might want to opt-out reliability, which is often the case for non-critical events such as particle effects, sounds effects etc.
+
+Reliability configuration can be specified for both `ServerRpc` and `ClientRpc` methods at compile time:
+
+```cs
+[ServerRpc]
+void MyReliableServerRpc() { /* ... */ }
+
+[ServerRpc(IsReliable = false)]
+void MyUnreliableServerRpc() { /* ... */ }
+
+[ClientRpc]
+void MyReliableClientRpc() { /* ... */ }
+
+[ClientRpc(IsReliable = false)]
+void MyUnreliableClientRpc() { /* ... */ }
+```
+
+An RPC call made without active connection will be dropped and will not be queued for send automatically. Both reliable and unreliable RPC calls have to be made when there is an active network connection established between a client and the server. Also reliable RPC calls made during connection will be dropped on disconnect as well.
+
+### Execution Table
+
+An RPC function **never** executes its body immediately since it's being a network construct. Even a `ServerRpc` called by a host (an instance that is a client and the server at the same time, aka listen server) will not be executed immediately but follow the regular network frame staging first.
+
+||Server|Client|Host (Server+Client)|
+|-:|:-:|:-:|:-:|
+|ServerRpc Network Send|❌|✅|✅|
+|ServerRpc Network Call|✅|❌|✅|
+|ServerRpc Direct Call|❌|❌|❌|
+|ClientRpc Network Send|✅|❌|✅|
+|ClientRpc Network Call|❌|✅|✅|
+|ClientRpc Direct Call|❌|❌|❌|
+
+## RPC Params
+
+// todo: descriptions needed below
+
```

Comment:
Unity's old legacy networking (and consequently, PUN, which is mentioned in the Prior Work section in the RFC) has [NetworkMessageInfo](https://docs.unity3d.com/560/Documentation/ScriptReference/NetworkMessageInfo.html) ([PhotonMessageInfo](https://doc-api.photonengine.com/en/pun/v2/struct_photon_1_1_pun_1_1_photon_message_info.html) in the case of PUN). Both include timestamps.

With a reliable NetworkTime (like [Network.time](https://docs.unity3d.com/560/Documentation/ScriptReference/Network-time.html)), this makes it comparatively easy to estimate the real position based on position when the RPC was sent and RPC transition time (and velocity). It's also useful for sanity checks on whether a given RPC is (still) valid.

This is useful for both, ClientRPC and ServerRPC and is currently missing in both. It depends on having reliable network time, of course.

Reply:
I think it'd make sense to have more context info in the `XXXReceiveParams`. This RFC only a future-proof way to add extra info later without breaking existing RPC API. One of those upcoming additional fields might well be network time, timestamp, frame number etc.